### ISSUE TITLE
momentum direction from GblFitter

### ISSUE THREAD

**Issue by IuriiSorokin** [GenFit/GenFit#73]
Dear Genfit developers,

I encountered the following behavior of GblFitter:
When fitting a Track with GblFitter, in case the DetPlane normal vector is opposite to the particle flight direction (as defined by the seed values), the fitted momentum on this plane has a wrong sign.

Is it a bug, or the expected behavior?

Here is a code that reproduces the case:
    https://github.com/IuriiSorokin/simple_gbl_fit/
There I fit a Track to five PlanarMeasurements. The PlanarMeasurements are defined on the DetPlanes:

  DetPlane( {0,0,10}, { 1, 0, 0}, {0, 1, 0} )
  DetPlane( {0,0,30}, {-1, 0, 0}, {0, 1, 0} )  <-- notice the u-vector
  DetPlane( {0,0,50}, { 1, 0, 0}, {0, 1, 0} )
  DetPlane( {0,0,70}, { 1, 0, 0}, {0, 1, 0} )
  DetPlane( {0,0,90}, { 1, 0, 0}, {0, 1, 0} )

Notice that the second plane is flipped around the y-axis.
The Track is seeded with the position {0, 0, 0}, and the momentum { 0, 0, 0.2 } (i.e. along the z-axis).
The true track position and momentum (used to create the PlanarMeasurements) are {0, 0, 0} and { -0.01, 0,  0.1 } respectively.

There is a uniform magnetic field { 0, 1, 0 }.

After fitting I get the following momenta on the DetPlanes, respectively:

  { 0.012301,-0.000022, 0.106242}
  {-0.006220, 0.000025,-0.107519}  <-- notice the sign
  { 0.000348,-0.000027, 0.108468}
  {-0.005769,-0.000029, 0.109111}
  {-0.012251,-0.000029, 0.109402}

The values are close to the true ones, except of the sign on the second plane.

I am using Genfit commit 70c408ff7424bdfd572ddc175b9673e8a1e91bbb from Jul 14, 2017 

I hope you can help me to clarify this issue.

Thank you in advance,
Iurii

**Comment by TadeasB** [comment]
Hello,

is this really specific to GBLFitter? It seems to me this is rather some improper use (or not using at all) of local <-> global transformation (see StateOnPlane and related). I would guess you ask for the state in local coordinates of the plane - then it is obvious you get opposite momenta for opposite local w-coordinate of the plane.

**Comment by IuriiSorokin** [comment]
Dear Tadeas,

Thank you for picking up the old issue.

I think it's specific to GBL. If I understand it correctly, in RK track representation (`q/p`, `t_u`, `t_v`, `u`, `v`) the track direction is undefined. The term `p` is the absolute momentum, and is always positive (also can't be 0). Class `RKTrackRep` stores the momentum direction in `StateOnPlane::auxInfo_`, and it can be accessed with `RKTrackRep::setSpu(...)`, and `RKTrackRep::getSpu(...)` methods. 

I think the transformation between local and global reference frames is done correctly in `RKTrackRep`. In the above example, to construct the measurements for the fit, I extrapolate the track from `{0,0,0}` to each of the planes. And I get correct momentum direction on all planes, including the flipped one. I also checked the `Spu`: as expected, it is positive on planes 0,2,3,4, and negative on the filpped plane.

**Comment by TadeasB** [comment]
Hello,

so can you please try with a different fitter?

Regards,
Tadeas

**Comment by IuriiSorokin** [comment]
Checked. It works correctly with `KalmanFilter`. Global momentum has a correct sign, and spu is negative on the flipped plane. Below I quote the output with the numbers. 

```
 === Constructing the world ===
Info in <TGeoManager::TGeoManager>: Geometry GenfitGeometry, GenfitGeometry created
Silicon radiation length = 9.35362
Vacuum radiation length = 1e+30
 === Constructing the measurements ===
Extrapolating to plane 0
Init pos: TVector3 A 3D physics vector (x,y,z)=(0.000000,0.000000,0.000000) (rho,theta,phi)=(0.000000,0.000000,0.000000)
Init mom: TVector3 A 3D physics vector (x,y,z)=(-0.010000,0.000000,0.100000) (rho,theta,phi)=(0.100499,5.710593,180.000000)
Ext. pos: TVector3 A 3D physics vector (x,y,z)=(-0.849023,0.000000,10.000000) (rho,theta,phi)=(10.035977,4.852907,180.000000)
Ext. mom: TVector3 A 3D physics vector (x,y,z)=(-0.006950,0.000000,0.099506) (rho,theta,phi)=(0.099749,3.995188,180.000000)
[simple_kf_fit.cxx:128] rep.getSpu( state )=1
Meas.pos: TVector3 A 3D physics vector (x,y,z)=(-0.848420,-0.000262,10.000000) (rho,theta,phi)=(10.035926,4.849475,-179.982312)
Meas.err: TVector3 A 3D physics vector (x,y,z)=(0.000603,-0.000262,0.000000) (rho,theta,phi)=(0.000658,90.000000,-23.469099)
Extrapolating to plane 1
Init pos: TVector3 A 3D physics vector (x,y,z)=(-0.849023,0.000000,10.000000) (rho,theta,phi)=(10.035977,4.852907,180.000000)
Init mom: TVector3 A 3D physics vector (x,y,z)=(-0.006950,0.000000,0.099506) (rho,theta,phi)=(0.099749,3.995188,180.000000)
Ext. pos: TVector3 A 3D physics vector (x,y,z)=(-1.637839,0.000000,30.000000) (rho,theta,phi)=(30.044675,3.124940,180.000000)
Ext. mom: TVector3 A 3D physics vector (x,y,z)=(-0.000895,-0.000000,0.098259) (rho,theta,phi)=(0.098263,0.521989,-180.000000)
[simple_kf_fit.cxx:128] rep.getSpu( state )=-1
Meas.pos: TVector3 A 3D physics vector (x,y,z)=(-1.169562,-0.017779,30.000000) (rho,theta,phi)=(30.022795,2.232826,-179.129066)
Meas.err: TVector3 A 3D physics vector (x,y,z)=(0.468277,-0.017779,0.000000) (rho,theta,phi)=(0.468614,90.000000,-2.174356)
Extrapolating to plane 2
Init pos: TVector3 A 3D physics vector (x,y,z)=(-1.637839,0.000000,30.000000) (rho,theta,phi)=(30.044675,3.124940,180.000000)
Init mom: TVector3 A 3D physics vector (x,y,z)=(-0.000895,-0.000000,0.098259) (rho,theta,phi)=(0.098263,0.521989,-180.000000)
Ext. pos: TVector3 A 3D physics vector (x,y,z)=(-1.204958,0.000000,50.000000) (rho,theta,phi)=(50.014517,1.380513,180.000000)
Ext. mom: TVector3 A 3D physics vector (x,y,z)=(0.005069,0.000000,0.096660) (rho,theta,phi)=(0.096793,3.001894,0.000000)
[simple_kf_fit.cxx:128] rep.getSpu( state )=1
Meas.pos: TVector3 A 3D physics vector (x,y,z)=(-0.065005,-0.075216,50.000000) (rho,theta,phi)=(50.000099,0.113920,-130.835225)
Meas.err: TVector3 A 3D physics vector (x,y,z)=(1.139953,-0.075216,0.000000) (rho,theta,phi)=(1.142432,90.000000,-3.774993)
Extrapolating to plane 3
Init pos: TVector3 A 3D physics vector (x,y,z)=(-1.204958,0.000000,50.000000) (rho,theta,phi)=(50.014517,1.380513,180.000000)
Init mom: TVector3 A 3D physics vector (x,y,z)=(0.005069,0.000000,0.096660) (rho,theta,phi)=(0.096793,3.001894,0.000000)
Ext. pos: TVector3 A 3D physics vector (x,y,z)=(0.473261,0.000000,70.000000) (rho,theta,phi)=(70.001600,0.387364,0.000000)
Ext. mom: TVector3 A 3D physics vector (x,y,z)=(0.010943,0.000000,0.094704) (rho,theta,phi)=(0.095334,6.591188,0.000000)
[simple_kf_fit.cxx:128] rep.getSpu( state )=1
Meas.pos: TVector3 A 3D physics vector (x,y,z)=(-1.716799,-0.166550,70.000000) (rho,theta,phi)=(70.021248,1.411530,-174.458965)
Meas.err: TVector3 A 3D physics vector (x,y,z)=(-2.190060,-0.166550,0.000000) (rho,theta,phi)=(2.196383,90.000000,-175.651134)
Extrapolating to plane 4
Init pos: TVector3 A 3D physics vector (x,y,z)=(0.473261,0.000000,70.000000) (rho,theta,phi)=(70.001600,0.387364,0.000000)
Init mom: TVector3 A 3D physics vector (x,y,z)=(0.010943,0.000000,0.094704) (rho,theta,phi)=(0.095334,6.591188,0.000000)
Ext. pos: TVector3 A 3D physics vector (x,y,z)=(3.436181,0.000000,90.000000) (rho,theta,phi)=(90.065572,2.186479,0.000000)
Ext. mom: TVector3 A 3D physics vector (x,y,z)=(0.016726,0.000000,0.092378) (rho,theta,phi)=(0.093880,10.262578,0.000000)
[simple_kf_fit.cxx:128] rep.getSpu( state )=1
Meas.pos: TVector3 A 3D physics vector (x,y,z)=(3.466203,-1.349709,90.000000) (rho,theta,phi)=(90.076835,2.366697,-21.275499)
Meas.err: TVector3 A 3D physics vector (x,y,z)=(0.030022,-1.349709,0.000000) (rho,theta,phi)=(1.350043,90.000000,-88.725749)
 === Fitting === 
 ================================== State 0 ============================================
genfit::MeasuredStateOnPlane my address 0x32eb430 my plane's address 0x2043900; use count: 11
 state vector: 
Vector (5)  is as follows

     |        1  |
------------------
   0 |-7.57641 
   1 |0.0747638 
   2 |-0.000387239 
   3 |-0.84842 
   4 |-0.000261923 

 covariance matrix: 
5x5 matrix is as follows

     |      0    |      1    |      2    |      3    |      4    |
----------------------------------------------------------------------
   0 |      44.76      0.1492   2.096e-05  -2.193e-07   1.698e-09 
   1 |     0.1492    0.001068   4.327e-07  -5.149e-09   3.139e-12 
   2 |  2.096e-05   4.327e-07   0.0005359  -2.749e-12  -4.257e-09 
   3 | -2.193e-07  -5.149e-09  -2.749e-12   9.109e-08   1.829e-17 
   4 |  1.698e-09   3.139e-12  -4.257e-09   1.829e-17   9.065e-08 

 defined in plane DetPlane: O(0, 0, 10) u(1, 0, 0) v(0, 1, 0) n(0, 0, 1) 
 3D position: TVector3 A 3D physics vector (x,y,z)=(-0.848420,-0.000262,10.000000) (rho,theta,phi)=(10.035926,4.849475,-179.982312)
 3D momentum: TVector3 A 3D physics vector (x,y,z)=(0.009841,-0.000051,0.131621) (rho,theta,phi)=(0.131989,4.275753,-0.296761)
[simple_kf_fit.cxx:178] dynamic_cast<const RKTrackRep*>(track.getFittedState(iState).getRep())->getSpu( track.getFittedState(iState) )=1
 ================================== State 1 ============================================
genfit::MeasuredStateOnPlane my address 0x32eb600 my plane's address 0x20439b0; use count: 10
 state vector: 
Vector (5)  is as follows

     |        1  |
------------------
   0 |-7.60555 
   1 |0.116273 
   2 |0.000449045 
   3 |-1.03385 
   4 |-0.0088975 

 covariance matrix: 
5x5 matrix is as follows

     |      0    |      1    |      2    |      3    |      4    |
----------------------------------------------------------------------
   0 |      42.29     -0.1115   6.722e-05     -0.2864  -0.0009267 
   1 |    -0.1115   0.0005516  -3.995e-07   -0.003101   6.234e-06 
   2 |  6.722e-05  -3.995e-07   0.0002415   3.299e-06   -0.003629 
   3 |    -0.2864   -0.003101   3.299e-06     0.07998  -5.706e-05 
   4 | -0.0009267   6.234e-06   -0.003629  -5.706e-05      0.0742 

 defined in plane DetPlane: O(0, 0, 30) u(-1, 0, 0) v(0, 1, 0) n(0, 0, -1) 
 3D position: TVector3 A 3D physics vector (x,y,z)=(1.033847,-0.008898,30.000000) (rho,theta,phi)=(30.017810,1.973795,-0.493087)
 3D momentum: TVector3 A 3D physics vector (x,y,z)=(0.015186,-0.000059,0.130603) (rho,theta,phi)=(0.131483,6.632212,-0.221275)
[simple_kf_fit.cxx:178] dynamic_cast<const RKTrackRep*>(track.getFittedState(iState).getRep())->getSpu( track.getFittedState(iState) )=-1
 ================================== State 2 ============================================
genfit::MeasuredStateOnPlane my address 0x32ec680 my plane's address 0x2043a60; use count: 9
 state vector: 
Vector (5)  is as follows

     |        1  |
------------------
   0 |9.01891 
   1 |-0.0130829 
   2 |-0.00506973 
   3 |1.06179 
   4 |-0.148942 

 covariance matrix: 
5x5 matrix is as follows

     |      0    |      1    |      2    |      3    |      4    |
----------------------------------------------------------------------
   0 |      9.606    -0.04604  -3.587e-05      0.1156   -0.001389 
   1 |   -0.04604    0.000424  -2.417e-07    0.003216  -3.274e-06 
   2 | -3.587e-05  -2.417e-07   0.0001729  -9.915e-06    0.003193 
   3 |     0.1156    0.003216  -9.915e-06      0.1494  -0.0002486 
   4 |  -0.001389  -3.274e-06    0.003193  -0.0002486      0.1323 

 defined in plane DetPlane: O(0, 0, 50) u(1, 0, 0) v(0, 1, 0) n(0, 0, 1) 
 3D position: TVector3 A 3D physics vector (x,y,z)=(1.061792,-0.148942,50.000000) (rho,theta,phi)=(50.011495,1.228448,-7.984992)
 3D momentum: TVector3 A 3D physics vector (x,y,z)=(-0.001450,-0.000562,0.110867) (rho,theta,phi)=(0.110878,0.803856,-158.818278)
[simple_kf_fit.cxx:178] dynamic_cast<const RKTrackRep*>(track.getFittedState(iState).getRep())->getSpu( track.getFittedState(iState) )=1
 ================================== State 3 ============================================
genfit::MeasuredStateOnPlane my address 0x32ec850 my plane's address 0x2043b10; use count: 10
 state vector: 
Vector (5)  is as follows

     |        1  |
------------------
   0 |9.55074 
   1 |-0.0214071 
   2 |-0.015113 
   3 |0.510444 
   4 |-0.336334 

 covariance matrix: 
5x5 matrix is as follows

     |      0    |      1    |      2    |      3    |      4    |
----------------------------------------------------------------------
   0 |      14.28    -0.07761   1.655e-05     -0.9935   0.0004735 
   1 |   -0.07761    0.001849   6.762e-08      0.0173  -5.323e-06 
   2 |  1.655e-05   6.762e-08    0.001094  -3.049e-06    0.007386 
   3 |    -0.9935      0.0173  -3.049e-06      0.5259  -5.907e-05 
   4 |  0.0004735  -5.323e-06    0.007386  -5.907e-05      0.3596 

 defined in plane DetPlane: O(0, 0, 70) u(1, 0, 0) v(0, 1, 0) n(0, 0, 1) 
 3D position: TVector3 A 3D physics vector (x,y,z)=(0.510444,-0.336334,70.000000) (rho,theta,phi)=(70.002669,0.500334,-33.381125)
 3D momentum: TVector3 A 3D physics vector (x,y,z)=(-0.002241,-0.001582,0.104668) (rho,theta,phi)=(0.104704,1.501053,-144.778667)
[simple_kf_fit.cxx:178] dynamic_cast<const RKTrackRep*>(track.getFittedState(iState).getRep())->getSpu( track.getFittedState(iState) )=1
 ================================== State 4 ============================================
genfit::MeasuredStateOnPlane my address 0x32eca20 my plane's address 0x2043bc0; use count: 11
 state vector: 
Vector (5)  is as follows

     |        1  |
------------------
   0 |9.9562 
   1 |-0.0556058 
   2 |-0.0175308 
   3 |-0.261452 
   4 |-0.675284 

 covariance matrix: 
5x5 matrix is as follows

     |      0    |      1    |      2    |      3    |      4    |
----------------------------------------------------------------------
   0 |      16.58     -0.1781  -9.482e-05      -3.329  -0.0003659 
   1 |    -0.1781    0.004719   3.146e-06     0.08981   2.942e-05 
   2 | -9.482e-05   3.146e-06    0.001949   4.969e-05     0.03438 
   3 |     -3.329     0.08981   4.969e-05        2.38   0.0004768 
   4 | -0.0003659   2.942e-05     0.03438   0.0004768       1.159 

 defined in plane DetPlane: O(0, 0, 90) u(1, 0, 0) v(0, 1, 0) n(0, 0, 1) 
 3D position: TVector3 A 3D physics vector (x,y,z)=(-0.261452,-0.675284,90.000000) (rho,theta,phi)=(90.002913,0.460986,-111.165069)
 3D momentum: TVector3 A 3D physics vector (x,y,z)=(-0.005576,-0.001758,0.100270) (rho,theta,phi)=(0.100440,3.336785,-162.501545)
[simple_kf_fit.cxx:178] dynamic_cast<const RKTrackRep*>(track.getFittedState(iState).getRep())->getSpu( track.getFittedState(iState) )=1
```

**Comment by TadeasB** [comment]
Hello,

interesting. Unfortunately I do not have any idea, what could be the reason :-(
Any ideas?

**Comment by TadeasB** [comment]
Seems the auxInfo is not copied over at some place - honestly I've never touched this...

**Comment by TadeasB** [comment]
OK... it seems this is happening in GblFItterInfo at several place, when new (Measured)StateOnPlane is created and the auxInfo is not passed from the old one...

As you have a test setup for this, would you volunteer to try to make a fix?

To do so, one would have on each occasion a StateOnPlane(state, cov, plane) or similar with MeasuredStateOnPlane also pass the auxInfo of the original state as the last argument.

**Comment by IuriiSorokin** [comment]
Ok, I'll try...

**Comment by IuriiSorokin** [comment]
I think the problem is rather in the fact that `RKTrackRep` state vector does not contain the direction information.

As far as I understand, `GblFitter` was designed with the idea to be agnostic about the underlying track representation. `GblFitter` and `GblFitterInfo` operate internally with the state vectors, and never fetch AuxInfo. 

Thus, in `GblFitterInfo` the fitted `MeasuredStateOnPlane` is constructed in:
https://github.com/GenFit/GenFit/blob/e9c2d5707839c927f882989a324486aee407c6ea/GBL/src/GblFitterInfo.cc#L86
and
https://github.com/GenFit/GenFit/blob/e9c2d5707839c927f882989a324486aee407c6ea/GBL/src/GblFitterInfo.cc#L402
The constructor 
```MeasuredStateOnPlane::MeasuredStateOnPlane(const TVectorD& state, const TMatrixDSym& cov, const SharedPlanePtr& plane, const AbsTrackRep* rep, const TVectorD& auxInfo)```
calls
```StateOnPlane::StateOnPlane(const TVectorD& state, const SharedPlanePtr& plane, const AbsTrackRep* rep)```
which initializes `auxInfo_` with an empty vector, and then `RKTrackRep::getSpu(const StateOnPlane& state)` returns the default `+1`.

I see two possible solutions here:
a) To change `RKTrackRep` such that the first element of the state vector would also contain the track direction.
b) Introduce a convention that track direction is stored separately (for all track representations, regardless of whether it is also contained in the state vector itself). I would then rather store it in a dedicated member variable of `StateOnPlane`.

**Comment by IuriiSorokin** [comment]
I also find it strange, that in the above example, `KalmanFilter` returns fitted states with negative `state[0]` on the first two planes, and positive on the other three.

**Comment by TadeasB** [comment]
When looking in the code, I see 
```

inline MeasuredStateOnPlane::MeasuredStateOnPlane(const TVectorD& state, const TMatrixDSym& cov, const SharedPlanePtr& plane, const AbsTrackRep* rep, const TVectorD& auxInfo) :
  StateOnPlane(state, plane, rep, auxInfo), cov_(cov)
{
  assert(rep != nullptr);
  //assert(cov_.GetNcols() == (signed)rep->getDim());
}
```
which seems correct. I think what I proposed should work. Let me do the modification to GblFItterInfo.cc and make a pull request. I will add you as reviewer - so you can check with your test setup. It should not harm anything anyway - this was my fault when writing this class quite some years ago... I was not aware of these subtle details. But with your explanation of how the track direction is book-kept, the fix seems obvious to me.

**Comment by TadeasB** [comment]
Can you please have a look into the new pull request - I cannot find a way to add you as a reviewer ... still looking for some button to do so :-)

**Comment by nils-braun** [comment]
You can not add him as a reviewer as he is not a "formal" genfit-repository member :-)
But no worry, you can still mention him in the PR and he will get mails (as you did).

**Comment by TadeasB** [comment]
@IuriiSorokin Thank you very much for reporting this issue and helping with the solution and testing. Sorry, it took so long for me to discover the issue report...

**Comment by TadeasB** [comment]
@nils-braun Can you close this issue? You probably have a button for this. Sorry that I did not do the magical " closes #73 " in the commit - I learned that only afterwards, how one can do this on github.
